Tradução: functions.qmd (#72)
cienciadedatos/pt-r4ds: functions.qmd

> # Funções {#sec-functions}

Reviewer: ```suggestion
# ✅ Funções {#sec-functions}
```

Neste capítulo, você aprenderá sobre três tipos úteis de funções:

-   Funções vetoriais que recebem um ou mais vetores como entrada e retornam um vetor como saída.
-   Funções de *data frames* que recebem um *data frame* como entrada e retornam um *data frame* como saída.
-   Funções de plotagem que recebem um *data frame* como entrada e retornam um gráfico como saída.

Cada uma dessas seções inclui muitos exemplos para ajudá-lo a generalizar esses padrões que você vê.
Esses exemplos não seriam possíveis sem a ajuda do pessoal do Twitter, e encorajamos você seguir os links nos comentários para ver as inspirações originais.
Você também pode querer ler os tweets motivacionais originais para [funções gerais](https://twitter.com/hadleywickham/status/1571603361350164486) e [funções de plotagem](https://twitter.com/hadleywickham/status/1574373127349575680) para ver ainda mais funções.

### Pré-requisitos

Resumiremos uma variedade de funções do pacote tidyverse.
Também usaremos o pacote dados como fonte de dados familiares para usar em nossas funções.

```{r}
#| message: false
library(tidyverse)
library(dados)
```

> ## Funções vetorias

Reviewer 2: ```suggestion
## Funções vetoriais
```

(d - min(d, na.rm = TRUE)) / (max(d, na.rm = TRUE) - min(d, na.rm = TRUE))  
```

> Para deixar isso um pouco mais claro, podemos substituir o bit que varia com `█`:

Reviewer 2: ```suggestion
Para deixar isso um pouco mais claro, podemos substituir a parte que varia com `█`:
```

```{r}
#| eval: false
(█ - min(█, na.rm = TRUE)) / (max(█, na.rm = TRUE) - min(█, na.rm = TRUE))
```

Para transformar isso em uma função você precisa de três coisas:

1.  Um **nome**.
    Aqui usaremos `rescala01` porque esta função redimensiona um vetor para ficar entre 0 e 1.

2.  Os **argumentos**.
    Os argumentos variam entre as chamadas e nossa análise acima nos diz que temos apenas um.
    Chamaremos de `x` porque este é o nome convencional para um vetor numérico.

3.  O **corpo**.
    O corpo é o código repetido em todas as chamadas.

Então, você cria uma função seguindo o modelo:

```{r}
nome <- function(argumentos) {
  corpo
}
```

Para este caso nos leva a:

```{r}
rescala01 <- function(x) {
  (x - min(x, na.rm = TRUE)) / (max(x, na.rm = TRUE) - min(x, na.rm = TRUE))
}
```

Neste ponto você pode testar com algumas entradas simples para ter certeza de que capturou a lógica corretamente:

```{r}
rescala01(c(-10, 0, 10))
rescala01(c(1, 2, 3, NA, 5))
```

Então você pode reescrever a chamada para `mutate()` como:

```{r}
df |> mutate(
  a = rescala01(a),
  b = rescala01(b),
  c = rescala01(c),
  d = rescala01(d),
)
```

> (No @sec-iteration, ocê aprenderá como usar `across()` para reduzir ainda mais a duplicação, então tudo que você precisará é `df |> mutate(across(a:d, rescala01))`).

Reviewer 2: ```suggestion
(No @sec-iteration, você aprenderá como usar `across()` para reduzir ainda mais a duplicação, então tudo que você precisará é `df |> mutate(across(a:d, rescala01))`).
```

### Melhorando nossa função

> Você pode notar que a função `rescala01()` faz algum trabalho desnecessário --- em vez de calcular `min()` duas vezes e `max()` uma vez que poderíamos calcular o mínimo e o máximo em uma única etapa com `range( )`:

Reviewer 2: ```suggestion
Você pode notar que a função `rescala01()` faz algum trabalho desnecessário --- em vez de calcular `min()` duas vezes e `max()` uma vez, poderíamos calcular o mínimo e o máximo em uma única etapa com `range( )`:
```

```{r}
prim_maiusculo <- function(x) {
  str_sub(x, 1, 1) <- str_to_upper(str_sub(x, 1, 1))
  x
}

prim_maiusculo("hello")
```

Ou talvez você queira retirar sinais de porcentagem, vírgulas e cifrões de uma string antes de convertê-la em um número:

```{r}
# https://twitter.com/NVlabormarket/status/1571939851922198530
limpa_numero <- function(x) {
  tem_pct <- str_detect(x, "%")
  num <- x |> 
    str_remove_all("%") |> 
    str_remove_all(",") |> 
    str_remove_all(fixed("$")) |> 
    as.numeric()
  if_else(tem_pct, num / 100, num)
}

limpa_numero("$12,300")
limpa_numero("45%")
```

Às vezes, suas funções serão altamente especializadas para uma etapa de análise de dados.
Por exemplo, se você tiver um monte de variáveis ​​que registram valores ausentes como 997, 998 ou 999, você pode querer escrever uma função para substituí-los por `NA`:

```{r}
ajusta_na <- function(x) {
  if_else(x %in% c(997, 998, 999), NA, x)
}
```

Nós nos concentramos em exemplos que usam um único vetor porque achamos que são os mais comuns.
Mas não há razão para que sua função não possa receber múltiplas vetores na entrada.

### Funções de resumo

> Outra família importante de funções vetoriais são as funções de resumo, funções que retornam um único valor para uso em `summarize()`.

Reviewer 2: No capítulo **Numbers** essas funções foram traduzidas como funções de sumarização, e aqui estão como funções de resumo.

Não sugeri a alteração pois são sinônimos, então acho que a diferença não prejudica o entendimento do leitor. Mas achei legal pontuar, talvez por questões de padronização seja interessante realizar as alterações e incluir o termo no guia de tradução.

> Outra família importante de funções vetoriais são as funções de resumo, funções que retornam um único valor para uso em `summarize()`.

Author: ```suggestion
Outra família importante de funções vetoriais são as funções de sumarização, funções que retornam um único valor para uso em `summarize()`.
```

    }
    ```

## Funções de *data frame*

Funções vetoriais são úteis para extrair código repetido em um verbo dplyr.
Mas muitas vezes você também repetirá os próprios verbos, especialmente em um *pipeline* grande.
Ao perceber que você está copiando e colando vários verbos dplyr diversas vezes, você pode pensar em escrever uma função de *data frame*.
As funções de *data frame* funcionam como verbos dplyr: elas pegam um *data frame* como primeiro argumento, alguns argumentos extras que dizem o que fazer com ele e retornam um *data frame* ou um vetor.

> Para permitir que você escreva uma função que use verbos dplyr, primeiro apresentaremos o desafio da indireção (*indirection challenge*) e como você pode superá-lo usando `{{ }}`.

Reviewer 2: ```suggestion
Para permitir que você escreva uma função que use verbos dplyr, primeiro apresentaremos o desafio da indireção (*indirection challenge*) e como você pode superá-lo usando os sinais de chaves `{{ }}`.
```

  y = 100
)

df |> media_agrupada(grupo, x)
df |> media_agrupada(grupo, y)
```

Independentemente de como chamamos `media_agrupada()` ele sempre faz `df |> group_by(var_grupo) |> summary(mean(var_media))`, em vez de `df |> group_by(grupo) |> summary(mean(x) )` ou `df |> group_by(grupo) |> summary(mean(y))`.
Este é um problema de indireção (*indirection*) e surge porque o dplyr usa **avaliação organizada** (*tidy evaluation*) para permitir que você se refira aos nomes das variáveis ​​dentro do seu *data frame* sem qualquer tratamento especial.

A avaliação organizada (*tidy evaluation*) é ótima 95% das vezes porque torna suas análises de dados muito concisas, já que você nunca precisa dizer de qual *data frame* vem uma variável; é óbvio pelo contexto.
A desvantagem da avaliação organizada surge quando queremos agrupar códigos repetidos do tidyverse em uma função.
Aqui precisamos de alguma maneira de dizer a `group_by()` e `summarize()` para não tratar `group_var` e `mean_var` como o nome das variáveis, mas em vez disso procurar dentro delas a variável que realmente queremos usar.

A avaliação organizada inclui uma solução para esse problema chamada **abraçar** (*embracing*) 🤗.

> Abraçar (*embracing*) uma variável significa envolvê-la entre colchetes para que (por exemplo) `var` se torne `{{ var }}`.

Reviewer 2: ```suggestion
Abraçar (*embracing*) uma variável significa envolvê-la entre chaves para que (por exemplo) `var` se torne `{{ var }}`.
```

Abraçar uma variável diz ao dplyr para usar o valor armazenado dentro do argumento, não o argumento como o nome literal da variável.

> Uma maneira de lembrar o que está acontecendo é pensar em `{{ }}` como olhar para dentro de um túnel --- `{{ var }}` fará uma função dplyr olhar dentro de `var` em vez de procurar por uma variável chamada ` var`.

Reviewer 3: ```suggestion
Uma maneira de lembrar o que está acontecendo é pensar em `{{ }}` como olhar para dentro de um túnel --- `{{ var }}` fará uma função dplyr olhar dentro de `var` em vez de procurar por uma variável chamada `var`.
```

Nas seções a seguir, exploraremos os tipos de funções úteis que você pode escrever depois de entender como abraçar (*embracing*) uma variável.

### Casos de uso mais comuns

Se você normalmente executa o mesmo conjunto de sumarização ao fazer a exploração inicial de dados, considere agrupá-los em uma função auxiliar:

```{r}
sumario6 <- function(data, var) {
  data |> summarize(
    min = min({{ var }}, na.rm = TRUE),
    media = mean({{ var }}, na.rm = TRUE),
    mediana = median({{ var }}, na.rm = TRUE),
    max = max({{ var }}, na.rm = TRUE),
    n = n(),
    n_faltantes = sum(is.na({{ var }})),
    .groups = "drop"
  )
}

diamante |> sumario6(quilate)
```

(Sempre que você usa `summarize()` em uma função auxiliar, achamos que é uma boa prática definir `.groups = "drop"` para evitar a mensagem de aviso e deixar os dados em um estado desagrupado.)

O bom dessa função é que, como ela usa `summarize()`, você pode usá-la em dados agrupados:

```{r}
diamante |> 
  group_by(corte) |> 
  sumario6(quilate)
```

> Além disso, uma vez que os argumentos da `summarize()` são usam mascamento de dados (*data-masking*), isso significa que o argumento `var` para a função `sumario6()` também é usa mascaramento de dados (*data-masking*).

Reviewer 2: ```suggestion
Além disso, uma vez que os argumentos da `summarize()` usam mascaramento de dados (*data-masking*), isso significa que o argumento `var` para a função `sumario6()` também usa mascaramento de dados (*data-masking*).
```

```{r}
unico_onde <- function(df, condicao, var) {
  df |> 
    filter({{ condicao }}) |> 
    distinct({{ var }}) |> 
    arrange({{ var }})
}

# Procura todos os destinos em Dezembro
voos |> unico_onde(mes == 12, destino)
```

Aqui abraçamos `condicao` porque é passado para `filter()` e `var` porque é passado para `distinct()` e `arrange()`.

Fizemos todos esses exemplos para usar um *data frame* como primeiro argumento, mas se você estiver trabalhando repetidamente com os mesmos dados, pode fazer sentido codificá-los.
Por exemplo, a função a seguir sempre funciona com o conjunto de dados de voos e sempre seleciona `data_hora`, `companhia_aerea` e `voo`, pois eles formam a chave primária composta (*compound primary key*) que permite identificar uma linha.

```{r}
subconjunto_voos <- function(linhas, colunas) {
  voos |> 
    filter({{ linhas }}) |> 
    select(data_hora, companhia_aerea, voo, {{ colunas }})
}
```

### Mascaramento de dados vs. seleção organizada

Às vezes você deseja selecionar variáveis ​​dentro de uma função que usa mascaramento de dados (*data-masking*).

> Por exemplo, imagine que você deseja escrever um `conta_faltantes()` que conte o número de observações faltantes nas linhas.

Reviewer 2: ```suggestion
Por exemplo, imagine que você deseja escrever uma função `conta_faltantes()` que conte o número de observações faltantes nas linhas.
```

```{r}
conta_faltantes <- function(df, grupo_vars, x_var) {
  df |> 
    group_by(pick({{ grupo_vars }})) |> 
    summarize(
      n_miss = sum(is.na({{ x_var }})),
      .groups = "drop"
  )
}

voos |> 
  conta_faltantes(c(ano, mes, dia), horario_saida)
```

Outro uso conveniente de `pick()` é fazer uma tabela 2d de contagens.
Aqui contamos usando todas as variáveis ​​nas `linhas` e `colunas`, então usamos `pivot_wider()` para reorganizar as contagens em uma grade:

```{r}
# https://twitter.com/pollicipes/status/1571606508944719876
conta_wide <- function(data, linhas, colunas) {
  data |> 
    count(pick(c({{ linhas }}, {{ colunas }}))) |> 
    pivot_wider(
      names_from = {{ colunas }}, 
      values_from = n,
      names_sort = TRUE,
      values_fill = 0
    )
}

diamante |> conta_wide(c(transparencia, cor), corte)
```

Embora nossos exemplos tenham se concentrado principalmente no dplyr, a avaliação organizada (*tidy-evaluation*) também e usada pelo pacote tidyr, e se você olhar a documentação `pivot_wider()` você pode ver que `names_from` usa seleção organizada (*tidy-selection*).

### Exercícios

> 1.  Usando os conjuntos de dados voos e clima do pacote dados, escreva uma função que:

Reviewer 2: ```suggestion
1.  Usando os conjuntos de dados "voos" e "clima" do pacote dados, escreva uma função que:
```

    1.  Encontra todos os voos que foram cancelados (ou seja, `is.na(horario_chegada)`) ou atrasados ​​por mais de uma hora.

        ```{r}
        #| eval: false
        voos |> filtra_problematicos()
        ```

>     2.  Conta o número de voos cancelados e o número de voos atrasados ​​mais de uma hora.

Reviewer 2: ```suggestion
    2.  Conta o número de voos cancelados e o número de voos atrasados por um tempo maior que uma hora.
```

        ```{r}
        #| eval: false
        voos |> group_by(destino) |> summariza_problematicos()
        ```

    3.  Encontra todos os voos que foram cancelados ou atrasados ​​por mais do que o número de horas fornecido pelo usuário:

        ```{r}
        #| eval: false
        voos |> filtra_problematicos(horas = 2)
        ```

    4.  Resuma o clima para calcular o mínimo, a média e o máximo de uma variável fornecida pelo usuário:

        ```{r}
        #| eval: false
        clima |> summariza(temperatura)
        ```

    5.  Converte a variável fornecida pelo usuário que usa o horário do relógio (por exemplo, `horario_saida`, `horario_chegada`, etc.) em um tempo decimal (ou seja, horas + (minutos/60)).

        ```{r}
        #| eval: false
        voos |> horario_padronizado(saida_programada)
        ```

> 2.  Para cada uma das funções a seguir, liste todos os argumentos que usam avaliação organizada (*tidy-evaluation) e descreva se eles usam mascaramento de dados (*data-masking*) ou seleção organizada (tidy-selection): `distinct()`, `count()`, `group_by()`, `rename_with()`, `slice_min()`, `slice_amostra()`.

Reviewer 3: ```suggestion
2.  Para cada uma das funções a seguir, liste todos os argumentos que usam avaliação organizada (*tidy-evaluation) e descreva se eles usam mascaramento de dados (*data-masking*) ou seleção organizada (tidy-selection): `distinct()`, `count()`, `group_by()`, `rename_with()`, `slice_min()`, `slice_sample()`.
```

valida_linearidade <- function(df, x, y) {
  df |>
    ggplot(aes(x = {{ x }}, y = {{ y }})) +
    geom_point() +
    geom_smooth(method = "loess", formula = y ~ x, color = "red", se = FALSE) +
    geom_smooth(method = "lm", formula = y ~ x, color = "blue", se = FALSE) 
}

dados_starwars |> 
  filter(massa < 1000) |> 
  valida_linearidade(massa, altura)
```

Ou talvez você queira uma alternativa aos gráficos de dispersão coloridos para conjuntos de dados muito grandes, onde a plotagem excessiva é um problema:

```{r}
#| fig-alt: |
#|   Gráfico hexagonal de preço versus quilate de diamantes mostrando uma relação positiva. 
#|   Existem mais diamantes com menos de 2 quilates do que com mais de 2 quilates.

# https://twitter.com/ppaxisa/status/1574398423175921665
grafico_hex <- function(df, x, y, z, bins = 20, fun = "mean") {
  df |> 
    ggplot(aes(x = {{ x }}, y = {{ y }}, z = {{ z }})) + 
    stat_summary_hex(
      aes(color = after_scale(fill)), # faz border igual ao preenchimento
      bins = bins, 
      fun = fun,
    )
}

diamante |> grafico_hex(quilate, preco, profundidade)
```

### Combinando com outros tidyverse

> Alguns das funções auxiliares mais úteis combinam uma pitada de manipulação de dados com ggplot2.

Reviewer 2: ```suggestion
Algumas das funções auxiliares mais úteis combinam uma pitada de manipulação de dados com ggplot2.
```

Por exemplo, se você quiser fazer um gráfico de barras verticais onde você classifica automaticamente as barras em ordem de frequência usando a função `fct_infreq()`.
Como o gráfico de barras é vertical, também precisamos inverter a ordem normal para obter os valores mais altos no topo:

```{r}
#| fig-alt: |

> #|   Gráfico de barras de transparência de diamantes, onde a clareza está no eixo y e conta 

Reviewer 2: ```suggestion
#|   Gráfico de barras de transparência de diamantes, onde a clareza está no eixo y e as contagens
```

Para contornar este problema, usamos o operador especial `:=` que a avaliação organizada (*tidy evaluation*) trata exatamente da mesma maneira que `=`..

Ou talvez você queira facilitar o desenho de um gráfico de barras apenas para um subconjunto de dados:

```{r}
#| fig-alt: |
#|   Gráfico de barras de clareza de diamantes. O mais comum é SI1, depois SI2 e depois 
#|   VS2, depois VS1, depois VVS2, depois VVS1, depois I1 e, por último, IF.

barras_condicionais <- function(df, condicao, var) {
  df |> 
    filter({{ condicao }}) |> 
    ggplot(aes(x = {{ var }})) + 
    geom_bar()
}

diamante |> barras_condicionais(corte == "Bom", transparencia)
```

Você também pode ser criativo e exibir resumos de dados de outras maneiras.
Você pode encontrar um aplicativo interessante em <https://gist.github.com/GShotwell/b19ef520b6d56f61a830fabb3454965b>; ele usa os rótulos dos eixos para exibir o valor mais alto.
À medida que você aprende mais sobre o ggplot2, o poder de suas funções continuará a aumentar.

Terminaremos com um caso mais complicado: rotular os gráficos que você cria.

### Rótulos

> Lembre-se da função de histograma que mostramos anteriormente?

Reviewer 2: ```suggestion
Lembra da função de histograma que mostramos anteriormente?
```

```{r}
histograma <- function(df, var, binwidth = NULL) {
  df |> 
    ggplot(aes(x = {{ var }})) + 
    geom_histogram(binwidth = binwidth)
}
```

> Não seria bom se pudéssemos rotular a saída com a variável e a largura do compartimento que foi usada?

Reviewer 2: ```suggestion
Não seria bom se pudéssemos rotular a saída com a variável e a largura do intervalo que foi usada?
```

### Exercícios

Crie uma função de plotagem rica implementando gradualmente cada uma das etapas abaixo:

1.  Desenhe um gráfico de dispersão dado o conjunto de dados e as variáveis ​​`x` e `y`.

2.  Adicione uma linha de melhor ajuste (*best fit*) (ou seja, um modelo linear sem erros padrão).

3.  Adicione um título.

## Estilos

R não se importa com o nome de sua função ou argumentos, mas os nomes fazem uma grande diferença para as pessoas.
Idealmente, o nome da sua função será curto, mas evocará claramente o que a função faz.
Isso é difícil!
Mas é melhor ser claro do que curto, pois o preenchimento automático do RStudio facilita a digitação de nomes longos.

Geralmente, os nomes das funções devem ser verbos e os argumentos devem ser substantivos.
Existem algumas exceções: substantivos são aceitáveis ​​se a função calcular um substantivo muito conhecido (ou seja, `media()` é melhor que `calcular_media()`), ou acessar alguma propriedade de um objeto (ou seja, `coef()` é melhor do que `obter_coeficientes()`).
Use seu bom senso e não tenha medo de renomear uma função se descobrir um nome melhor mais tarde.

```{r}
#| eval: false

# Muito curto
f()

> # Não pe verbo, nem descritivo

Reviewer 2: ```suggestion
# Não é verbo, nem descritivo
```

minha_funcao_maravilhosa()

# Longo, mas claro
adicionar_faltantes()
agrupar_anos()
```

R também não se importa em como você usa o espaço em branco em suas funções, mas os futuros leitores o farão.
Continue seguindo as regras do @sec-workflow-style.

> Além disso, `funcao()` deve sempre ser seguida por colchetes (`{}`), e o conteúdo deve ser recuado por dois espaços adicionais.

Reviewer 2: ```suggestion
Além disso, `funcao()` deve sempre ser seguida por chaves (`{}`), e o conteúdo deve ser recuado por dois espaços adicionais.
```
